readium/lcp/src/main/java/org/readium/r2/lcp/LcpContentProtection.kt

```diff
@@ -7,94 +7,192 @@
 package org.readium.r2.lcp
 
 import org.readium.r2.lcp.auth.LcpPassphraseAuthentication
-import org.readium.r2.lcp.service.LcpLicensedAsset
+import org.readium.r2.lcp.license.model.LicenseDocument
+import org.readium.r2.shared.asset.Asset
+import org.readium.r2.shared.error.ThrowableError
+import org.readium.r2.shared.error.Try
+import org.readium.r2.shared.error.getOrElse
+import org.readium.r2.shared.fetcher.ContainerFetcher
 import org.readium.r2.shared.fetcher.TransformingFetcher
-import org.readium.r2.shared.publication.ContentProtection
 import org.readium.r2.shared.publication.Publication
-import org.readium.r2.shared.publication.asset.FileAsset
-import org.readium.r2.shared.publication.asset.PublicationAsset
-import org.readium.r2.shared.publication.asset.RemoteAsset
+import org.readium.r2.shared.publication.protection.ContentProtection
 import org.readium.r2.shared.publication.services.contentProtectionServiceFactory
-import org.readium.r2.shared.util.Try
+import org.readium.r2.shared.resource.ArchiveFactory
+import org.readium.r2.shared.resource.Resource
+import org.readium.r2.shared.resource.ResourceFactory
+import org.readium.r2.shared.util.Url
+import org.readium.r2.shared.util.mediatype.MediaTypeRetriever
 
 internal class LcpContentProtection(
     private val lcpService: LcpService,
-    private val authentication: LcpAuthenticating
+    private val authentication: LcpAuthenticating,
+    private val mediaTypeRetriever: MediaTypeRetriever,
+    private val resourceFactory: ResourceFactory,
+    private val archiveFactory: ArchiveFactory
 ) : ContentProtection {
 
+    override val scheme: ContentProtection.Scheme =
+        ContentProtection.Scheme.Lcp
+
+    override suspend fun supports(
+        asset: Asset
+    ): Boolean =
+        lcpService.isLcpProtected(asset)
+
     override suspend fun open(
-        asset: PublicationAsset,
+        asset: Asset,
+        drmScheme: String,
+        credentials: String?,
+        allowUserInteraction: Boolean,
+        sender: Any?
+    ): Try<ContentProtection.Asset, Publication.OpeningException>? {
+        if (drmScheme != scheme.uri) {
+            return null
+        }
+
+        return when (asset) {
+            is Asset.Container -> openPublication(asset, credentials, allowUserInteraction, sender)
+            is Asset.Resource -> openLicense(asset, credentials, allowUserInteraction, sender)
+        }
```

Comment:
I know that it is not completely obvious from the architecture, but this is not supposed to happen. The `AssetRetriever` is the only way you should get an asset and it is responsible for transforming resources into containers in case of archives.
Actually, it's slightly visible through the `Asset.assetType` property. It's forbidden to be a `Resource` if you have the `Archive` `AssetType`.